### PR TITLE
Loki: remove `subqueries` from metrics.go logging and replace it with separate split and shard counters

### PR DESCRIPTION
**Pull request by slim-bean**
**What this PR does / why we need it**:

Currently the `metrics.go` log line emitted after every query includes a metric called "subqueries".  This currently tracks the number of queries created by the split_by_time operations done in Loki, but does not include any counts for subqueries created as a result of sharding.

It becomes difficult to make a single subqueries counter that gives useful information to determine how much a query is split by time and sharded by a shard factor, especially now that sharding in TSDB indexes is dynamic.

This PR removes and deprecates the `subqueries` stat and instead creates a `splits` and `shards` statistic which records how much a query was split_by_time and the total number of shards created as well.

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [x] Tests updated
- [x] `CHANGELOG.md` updated
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

File: pkg/querier/queryrange/codec_test.go [hunk 1]
```diff
@@ -1010,7 +1012,9 @@ var (
 			"execTime": 22,
 			"linesProcessedPerSecond": 23,
 			"queueTime": 21,
-			"subqueries": 1,
+			"shards": 0,
+            "splits": 0,
+			"subqueries": 0,
```
Comment (kavirajk) [inline pkg/querier/queryrange/codec_test.go:1017]: 1. `splits` is not formatted correctly looks like.
2. I would assume there shouldn't be `subqueries` keys here.

File: pkg/querier/queryrange/prometheus_test.go [hunk 2]
```diff
@@ -81,6 +81,8 @@ var emptyStats = `"stats": {
 		"execTime": 0,
 		"linesProcessedPerSecond": 0,
 		"queueTime": 0,
+        "splits": 0,
```
Comment (kavirajk) [inline pkg/querier/queryrange/prometheus_test.go:84]: formatting again.

File: pkg/logql/metrics_test.go [hunk 3]
```diff
@@ -111,7 +111,7 @@ func TestLogLabelsQuery(t *testing.T) {
 	})
 	require.Equal(t,
 		fmt.Sprintf(
-			"level=info org_id=foo traceID=%s latency=slow query_type=labels length=1h0m0s duration=25.25s status=200 label=foo throughput=100kB total_bytes=100kB total_entries=12\n",
+			"level=info org_id=foo traceID=%s latency=slow query_type=labels length=1h0m0s duration=25.25s status=200 label=foo splits=0 throughput=100kB total_bytes=100kB total_entries=12\n",
```
Comment (kavirajk) [inline pkg/logql/metrics_test.go:114]: I see we have test only for LabelQuery and SeriesQuery here. Good to add a test for `InstantQuery` as well? to see both `splits` and `shard` appear?

**Review by kavirajk — COMMENTED** [review]
Nice. thanks for adding this. LGTM with few minor comments.

**Comment by kavirajk** [comment]
wondering this may break existing dashboards based on `subqueries`. maybe add an entry in upgrade guide as well?

File: pkg/logql/metrics_test.go (hunk 3 above)
Comment (slim-bean) [reply pkg/logql/metrics_test.go:114]: I'm not entirely sure what these tests are actually testing? I just updated it to it wasn't failing but it seems strange to me to have a test that's validating the output of a log line?

I can see some arguments for this but we don't really have any contracts on log lines so I'm not sure we really need these tests?

File: pkg/querier/queryrange/codec_test.go (hunk 1 above)
Comment (slim-bean) [reply pkg/querier/queryrange/codec_test.go:1017]: subqueries still exists because it is deprecated but not removed.

**Comment by slim-bean** [comment]
> wondering this may break existing dashboards based on `subqueries`. maybe add an entry in upgrade guide as well?

added!

File: pkg/logql/metrics_test.go (hunk 3 above)
Comment (MasslessParticle) [reply pkg/logql/metrics_test.go:114]: I'd say that we should probably never test log lines

**Review by MasslessParticle — APPROVED** [review]

**Pull request merged**